File: packages/js/onnx2gguf/src/reader.ts
import { GGUFValueType, GGUFTensorType } from './builder';

export class GGUFReader {
  public kvs: Record<string, any> = {};
  public tensors: Record<
    string,
    { name: string; shape: bigint[]; type: GGUFTensorType; offset: bigint }
  > = {};
  public dataStart: number = 0;
  private view: DataView;
  private offset: number = 0;

  constructor(public buffer: ArrayBuffer) {
    this.view = new DataView(buffer);
    this.readHeader();
  }

  private readString(): string {
    const len = Number(this.view.getBigUint64(this.offset, true));
    this.offset += 8;
    const decoder = new TextDecoder();
    const str = decoder.decode(new Uint8Array(this.buffer, this.offset, len));
    this.offset += len;
    return str;
  }

  private readVal(vtype: GGUFValueType): any {
    switch (vtype) {
      case GGUFValueType.UINT8: {
        const v = this.view.getUint8(this.offset);
        this.offset += 1;
        return v;
      }
      case GGUFValueType.INT8: {
        const v = this.view.getInt8(this.offset);
        this.offset += 1;
        return v;
      }
      case GGUFValueType.UINT16: {
        const v = this.view.getUint16(this.offset, true);
        this.offset += 2;
        return v;
      }
      case GGUFValueType.INT16: {
        const v = this.view.getInt16(this.offset, true);
        this.offset += 2;
        return v;
      }
      case GGUFValueType.UINT32: {
        const v = this.view.getUint32(this.offset, true);
        this.offset += 4;
        return v;
      }
      case GGUFValueType.INT32: {
        const v = this.view.getInt32(this.offset, true);
        this.offset += 4;
        return v;
      }
      case GGUFValueType.FLOAT32: {
        const v = this.view.getFloat32(this.offset, true);
        this.offset += 4;
        return v;
      }
      case GGUFValueType.UINT64: {
        const v = this.view.getBigUint64(this.offset, true);
        this.offset += 8;
        return v;
      }
      case GGUFValueType.INT64: {
        const v = this.view.getBigInt64(this.offset, true);
        this.offset += 8;
        return v;
      }
      case GGUFValueType.FLOAT64: {
        const v = this.view.getFloat64(this.offset, true);
        this.offset += 8;
        return v;
      }
      case GGUFValueType.BOOL: {
        const v = this.view.getUint8(this.offset) !== 0;
        this.offset += 1;
        return v;
      }
      case GGUFValueType.STRING:
        return this.readString();
      case GGUFValueType.ARRAY: {
        const atype = this.view.getUint32(this.offset, true);
        this.offset += 4;
        const len = Number(this.view.getBigUint64(this.offset, true));
        this.offset += 8;
        const arr = [];
        for (let i = 0; i < len; i++) {
          arr.push(this.readVal(atype));
        }
        return arr;
      }
      default:
        throw new Error(`Unknown value type: ${vtype}`);
    }
  }

  private readHeader() {
    if (
      this.view.getUint8(0) !== 0x47 ||
      this.view.getUint8(1) !== 0x47 ||
      this.view.getUint8(2) !== 0x55 ||
      this.view.getUint8(3) !== 0x46
    ) {
      throw new Error('Not a GGUF file');
    }
    this.offset += 4;

    const version = this.view.getUint32(this.offset, true);
    this.offset += 4;
    if (version !== 2 && version !== 3) {
      throw new Error(`Unsupported GGUF version ${version}`);
    }

    const tensorCount = Number(this.view.getBigUint64(this.offset, true));
    this.offset += 8;
    const kvCount = Number(this.view.getBigUint64(this.offset, true));
    this.offset += 8;

    for (let i = 0; i < kvCount; i++) {
      const key = this.readString();
      const vtype = this.view.getUint32(this.offset, true);
      this.offset += 4;
      this.kvs[key] = this.readVal(vtype);
    }

    for (let i = 0; i < tensorCount; i++) {
      const name = this.readString();
      const nDims = this.view.getUint32(this.offset, true);
      this.offset += 4;
      const shape = [];
      for (let d = 0; d < nDims; d++) {
        shape.push(this.view.getBigUint64(this.offset, true));
        this.offset += 8;
      }
      const ttype = this.view.getUint32(this.offset, true);
      this.offset += 4;
      const offset = this.view.getBigUint64(this.offset, true);
      this.offset += 8;

      this.tensors[name] = { name, shape, type: ttype, offset };
    }

    const alignment = this.kvs['general.alignment'] || 32;
    const padding = (alignment - (this.offset % alignment)) % alignment;
    this.offset += padding;
    this.dataStart = this.offset;
  }

  public getTensor(name: string): Uint8Array {
    if (!this.tensors[name]) throw new Error(`Tensor ${name} not found`);
    const t = this.tensors[name];

    let items = 1n;
    for (const d of t.shape) items *= d;

    let size = 0n;
    if (t.type === GGUFTensorType.F32) size = items * 4n;
    else if (t.type === GGUFTensorType.F16) size = items * 2n;
    else if (t.type === GGUFTensorType.Q4_0) size = (items / 32n) * 18n;
    else if (t.type === GGUFTensorType.Q4_1) size = (items / 32n) * 20n;
    else if (t.type === GGUFTensorType.Q8_0) size = (items / 32n) * 34n;
    else throw new Error('Unknown type');

    const byteOffset = this.dataStart + Number(t.offset);
    return new Uint8Array(this.buffer, byteOffset, Number(size));
  }
}

File: packages/python/onnx9000-onnx2gguf/src/onnx9000/onnx2gguf/reader.py
"""Module providing onnx2gguf functionality."""

import struct
from typing import BinaryIO, Any
from .builder import GGUFValueType, GGUFTensorType


class GGUFReader:
    """Provides functionality for G G U F Reader."""

    def __init__(self, f: BinaryIO):
        """Initializes a new instance of the class."""
        self.f = f
        self.kvs = {}
        self.tensors = {}
        self._read_header()

    def _read_string(self) -> str:
        """read string."""
        length = struct.unpack("<Q", self.f.read(8))[0]
        return self.f.read(length).decode("utf-8")

    def _read_val(self, vtype: GGUFValueType) -> Any:
        """read val."""
        if vtype == GGUFValueType.UINT8:
            return struct.unpack("<B", self.f.read(1))[0]
        elif vtype == GGUFValueType.INT8:
            return struct.unpack("<b", self.f.read(1))[0]
        elif vtype == GGUFValueType.UINT16:
            return struct.unpack("<H", self.f.read(2))[0]
        elif vtype == GGUFValueType.INT16:
            return struct.unpack("<h", self.f.read(2))[0]
        elif vtype == GGUFValueType.UINT32:
            return struct.unpack("<I", self.f.read(4))[0]
        elif vtype == GGUFValueType.INT32:
            return struct.unpack("<i", self.f.read(4))[0]
        elif vtype == GGUFValueType.FLOAT32:
            return struct.unpack("<f", self.f.read(4))[0]
        elif vtype == GGUFValueType.UINT64:
            return struct.unpack("<Q", self.f.read(8))[0]
        elif vtype == GGUFValueType.INT64:
            return struct.unpack("<q", self.f.read(8))[0]
        elif vtype == GGUFValueType.FLOAT64:
            return struct.unpack("<d", self.f.read(8))[0]
        elif vtype == GGUFValueType.BOOL:
            return struct.unpack("<?", self.f.read(1))[0]
        elif vtype == GGUFValueType.STRING:
            return self._read_string()
        elif vtype == GGUFValueType.ARRAY:
            atype_int = struct.unpack("<I", self.f.read(4))[0]
            atype = GGUFValueType(atype_int)
            length = struct.unpack("<Q", self.f.read(8))[0]
            arr = []
            for _ in range(length):
                arr.append(self._read_val(atype))
            return arr
        else:
            raise ValueError(f"Unknown value type: {vtype}")

    def _read_header(self):
        """read header."""
        magic = self.f.read(4)
        if magic != b"GGUF":
            raise ValueError("Not a GGUF file")
        version = struct.unpack("<I", self.f.read(4))[0]
        if version not in (2, 3):
            raise ValueError(f"Unsupported GGUF version {version}")
        tensor_count = struct.unpack("<Q", self.f.read(8))[0]
        kv_count = struct.unpack("<Q", self.f.read(8))[0]
        for _ in range(kv_count):
            key = self._read_string()
            vtype_int = struct.unpack("<I", self.f.read(4))[0]
            vtype = GGUFValueType(vtype_int)
            val = self._read_val(vtype)
            self.kvs[key] = val
        for _ in range(tensor_count):
            name = self._read_string()
            n_dims = struct.unpack("<I", self.f.read(4))[0]
            shape = []
            for _ in range(n_dims):
                shape.append(struct.unpack("<Q", self.f.read(8))[0])
            ttype_int = struct.unpack("<I", self.f.read(4))[0]
            ttype = GGUFTensorType(ttype_int)
            offset = struct.unpack("<Q", self.f.read(8))[0]
            self.tensors[name] = {"name": name, "shape": shape, "type": ttype, "offset": offset}
        self.alignment = self.kvs.get("general.alignment", 32)
        pos = self.f.tell()
        padding = (self.alignment - pos % self.alignment) % self.alignment
        self.f.read(padding)
        self.data_start = self.f.tell()

    def get_tensor(self, name: str) -> bytes:
        """Gets tensor."""
        if name not in self.tensors:
            raise KeyError(name)
        t = self.tensors[name]
        items = 1
        for d in t["shape"]:
            items *= d
        dtype = t["type"]
        if dtype == GGUFTensorType.F32:
            size = items * 4
        elif dtype == GGUFTensorType.F16:
            size = items * 2
        elif dtype == GGUFTensorType.Q4_0:
            size = items // 32 * 18
        elif dtype == GGUFTensorType.Q4_1:
            size = items // 32 * 20
        elif dtype == GGUFTensorType.Q8_0:
            size = items // 32 * 34
        else:
            raise ValueError("Unknown type")
        self.f.seek(self.data_start + t["offset"])
        return self.f.read(size)
